lib/secrets/src/Cardano/Wallet/Primitive/Passphrase/Legacy.hs

```diff
@@ -63,34 +63,89 @@ import qualified Codec.CBOR.Write as CBOR
 import qualified Crypto.KDF.Scrypt as Scrypt
 import qualified Data.ByteArray as BA
 import qualified Data.ByteString.Char8 as B8
-import Data.Either.Extra
-    ( eitherToMaybe
-    )
 
 #if HAVE_SCRYPT
 import Crypto.Scrypt
     ( EncryptedPass (..)
     , Pass (..)
+    , Salt (Salt)
+    , encryptPass
+    , scryptParamsLen
     , verifyPass'
     )
 
--- | Verify a wallet spending password using the legacy Byron scrypt encryption
--- scheme.
-checkPassphrase :: Passphrase "encryption" -> PassphraseHash -> Maybe Bool
-checkPassphrase pwd stored = Just $
-    verifyPass' (Pass (BA.convert (cborify pwd))) encryptedPass
-  where
-    encryptedPass = EncryptedPass (BA.convert stored)
-
 haveScrypt :: Bool
 haveScrypt = True
+
 #else
--- | Stub function for when compiled without @scrypt@.
-checkPassphrase :: Passphrase "encryption" -> PassphraseHash -> Maybe Bool
-checkPassphrase _ _ = Nothing
 
 haveScrypt :: Bool
 haveScrypt = False
+
+#endif
+
+{- NOTE [LegacyScrypt]
+
+We need to transition away from the unmaintained `scrypt` package.
+Specifically, the `scrypt` package is not supported on `aarch64-darwin`.
+
+We can use the `cryptonite` package instead.
+In order to ensure that the new package produces the same result as the
+old one, we proceed as follows:
+
+    * in production:
+        * use `scrypt` package if available
+        * use `cryptonite` on `aarch64-darwin`
+    * in testing:
+        * generate random passphrases,
+            encrypted with `scrypt` package if available.
+        * check that these encrypted passphrases
+            are verified correctly with the `cryptonite` package
+
+These tests ensure that the code using `cryptonite` can verify
+hashed passphrases that were created with `scrypt`.
+
+-}
+
+{-----------------------------------------------------------------------------
+    Passphrase hashing and verification using the Scrypt algorithm
+------------------------------------------------------------------------------}
+-- | Verify a wallet spending password using the legacy Byron scrypt encryption
+-- scheme.
+checkPassphrase :: Passphrase "encryption" -> PassphraseHash -> Bool
+checkPassphrase = checkPassphraseCryptonite
+
+-- | Encrypt a wallet spending password using
+-- the legacy Byron scrypt encryption scheme.
+--
+-- Do not use this function in production, only in unit tests!
+encryptPassphraseTestingOnly
+    :: MonadRandom m
+    => PassphraseHashLength
+    -> Passphrase "encryption"
+    -> m PassphraseHash
+#if HAVE_SCRYPT
```

Comment:
did you take care of instantiating HAVE_SCRIPT variable in nix and linux/wndows CI environments?

Reply:
The variable `HAVE_SCRIPT` is set through the `scrypt` flag in the .cabal file.

https://github.com/cardano-foundation/cardano-wallet/blob/f265962e1f34d069854656b4836320e449a395ae/lib/secrets/cardano-wallet-secrets.cabal#L36-L46

This flag is `True` by default, but switched to `False` for `aarch64-darwin` systems in `./nix/haskell.nix`.